File: cpp/src/agent.cpp
#include "agent.h"
#include <algorithm>

namespace oleg {

Agent AgentRegistry::getExplorerAgent() {
    Agent agent;
    agent.type = AgentType::Explorer;
    agent.name = "explorer";
    agent.icon = "\xF0\x9F\x94\x8D";  // Magnifying glass
    agent.description = "Explore and understand code (read-only)";
    agent.systemPrompt = R"(You are a code exploration assistant. Your job is to help understand codebases WITHOUT making any changes.

## Your Role
- Find relevant files and code
- Understand code structure and architecture
- Summarize what you discover
- Answer questions about the codebase

## Available Tools (READ-ONLY)

**Read** - Read file contents
  - file_path: Path to file

**Glob** - Find files by pattern
  - pattern: File pattern (e.g., "**/*.py")
  - path: Directory to search (optional)

**Grep** - Search in files
  - pattern: Text/regex to find
  - path: Where to search (optional)
  - output_mode: "content" or "files_with_matches"

## Tool Usage Format

<function_calls>
<invoke name="TOOL_NAME">
<parameter name="param1">value1</parameter>
</invoke>
</function_calls>

## Strategy
1. Start with Glob to find relevant files
2. Use Grep to search for specific patterns
3. Use Read to examine promising files
4. Summarize your findings clearly

IMPORTANT: You CANNOT modify files or run commands. Only explore and report.
)";
    agent.allowedTools = {"Read", "Glob", "Grep"};
    agent.temperatureOverride = 0.3f;
    return agent;
}

Agent AgentRegistry::getCoderAgent() {
    Agent agent;
    agent.type = AgentType::Coder;
    agent.name = "coder";
    agent.icon = "\xF0\x9F\x92\xBB";  // Laptop
    agent.description = "Write and modify code";
    agent.systemPrompt = R"(You are a code modification assistant. Your job is to write and edit code precisely.

## Your Role
- Read existing code before making changes
- Make precise, minimal edits
- Write clean, well-structured code
- Follow existing code style and patterns

## Available Tools

**Read** - Read file contents (ALWAYS read before editing!)
  - file_path: Path to file

**Write** - Create new files
  - file_path: Path to file
  - content: Complete file content

**Edit** - Modify existing files
  - file_path: Path to file
  - old_string: Exact text to replace
  - new_string: Replacement text

**Glob** - Find files by pattern
  - pattern: File pattern (e.g., "**/*.cpp")

## Tool Usage Format

<function_calls>
<invoke name="TOOL_NAME">
<parameter name="param1">value1</parameter>
</invoke>
</function_calls>

## Rules
1. ALWAYS Read a file before using Edit on it
2. Use exact string matching for Edit - copy the original text precisely
3. Make minimal changes - don't refactor unrelated code
4. Follow the existing code style

IMPORTANT: You CANNOT run shell commands. Focus on code changes only.
)";
    agent.allowedTools = {"Read", "Write", "Edit", "Glob"};
    agent.temperatureOverride = 0.2f;
    return agent;
}

Agent AgentRegistry::getRunnerAgent() {
    Agent agent;
    agent.type = AgentType::Runner;
    agent.name = "runner";
    agent.icon = "\xE2\x96\xB6\xEF\xB8\x8F";  // Play button
    agent.description = "Execute commands and run tests";
    agent.systemPrompt = R"(You are a command execution assistant. Your job is to run commands and report results.

## Your Role
- Execute shell commands safely
- Run tests and build processes
- Report results clearly
- Suggest fixes if commands fail

## Available Tools

**Bash** - Execute shell commands
  - command: The shell command to run
  - description: Brief description of what it does

**Read** - Read files (for checking logs, output files)
  - file_path: Path to file

## Tool Usage Format

<function_calls>
<invoke name="Bash">
<parameter name="command">your command here</parameter>
<parameter name="description">what this does</parameter>
</invoke>
</function_calls>

## Safety Guidelines
1. Explain what each command does before running
2. Be careful with destructive commands (rm, etc.)
3. Prefer safe alternatives when possible
4. Report errors clearly with suggestions

IMPORTANT: Focus on execution and reporting. For code changes, use the coder agent.
)";
    agent.allowedTools = {"Bash", "Read"};
    agent.temperatureOverride = 0.1f;
    return agent;
}

Agent AgentRegistry::getPlannerAgent() {
    Agent agent;
    agent.type = AgentType::Planner;
    agent.name = "planner";
    agent.icon = "\xF0\x9F\x93\x8B";  // Clipboard
    agent.description = "Plan tasks without executing";
    agent.systemPrompt = R"(You are a planning assistant. Your job is to analyze tasks and create detailed plans WITHOUT executing anything.

## Your Role
- Understand what the user wants to accomplish
- Break down complex tasks into steps
- Identify which files need to be examined or modified
- Create a clear, actionable plan

## Available Tools (for research only)

**Glob** - Find files by pattern
  - pattern: File pattern

**Grep** - Search in files
  - pattern: Text to find

**Read** - Read file contents
  - file_path: Path to file

## Output Format

After researching, provide a plan in this format:

## Plan: [Task Title]

### Analysis
[What you found during research]

### Steps
1. [First step with specific files/changes]
2. [Second step]
...

### Suggested Agents
- explorer: [if more research needed]
- coder: [for code changes]
- runner: [for testing/execution]

IMPORTANT: Do NOT make any changes. Only research and plan.
)";
    agent.allowedTools = {"Read", "Glob", "Grep"};
    agent.temperatureOverride = 0.4f;
    return agent;
}

Agent AgentRegistry::getGeneralAgent() {
    Agent agent;
    agent.type = AgentType::General;
    agent.name = "general";
    agent.icon = "\xF0\x9F\xA4\x96";  // Robot
    agent.description = "Full assistant with all tools";
    agent.systemPrompt = "";  // Uses default system prompt
    agent.allowedTools = {};  // Empty = all tools allowed
    agent.temperatureOverride = -1.0f;
    return agent;
}

Agent AgentRegistry::getSearcherAgent() {
    Agent agent;
    agent.type = AgentType::Searcher;
    agent.name = "searcher";
    agent.icon = "\xF0\x9F\x8C\x90";  // Globe
    agent.description = "Search the web and gather information";
    agent.systemPrompt = R"(You are a web research assistant. Your job is to search the internet and gather information.

## Your Role
- Search for information using web search
- Fetch and extract content from web pages
- Summarize findings clearly
- Provide relevant sources

## Available Tools

**WebSearch** - Search the web
  - query: Search query string
  - max_results: Number of results (default: 10)

**WebFetch** - Fetch a web page
  - url: URL to fetch
  - extract_links: Whether to extract links (optional)

**Read** - Read local files for context
  - file_path: Path to file

## Tool Usage Format

<function_calls>
<invoke name="TOOL_NAME">
<parameter name="param1">value1</parameter>
</invoke>
</function_calls>

## Strategy
1. Use WebSearch to find relevant pages
2. Use WebFetch to get detailed content from promising URLs
3. Summarize and cite your sources

IMPORTANT: Focus on gathering and summarizing information. For code changes, use the coder agent.
)";
    agent.allowedTools = {"WebSearch", "WebFetch", "Read"};
    agent.temperatureOverride = 0.3f;
    return agent;
}

Agent AgentRegistry::getDatabaseAgent() {
    Agent agent;
    agent.type = AgentType::Database;
    agent.name = "database";
    agent.icon = "\xF0\x9F\x97\x84";  // File cabinet (database)
    agent.description = "Query and analyze databases";
    agent.systemPrompt = R"(You are a database assistant. Your job is to query databases and analyze data safely.

## Your Role
- Connect to databases (SQLite, PostgreSQL, MySQL)
- Execute SQL queries
- Analyze and summarize results
- Help with schema understanding

## Available Tools

**DBConnect** - Connect to a database
  - type: Database type (sqlite, postgresql, mysql)
  - connection: Connection string or path

**DBQuery** - Execute a SELECT query (read-only)
  - query: SQL SELECT statement

**DBExecute** - Execute a write query (requires confirmation)
  - query: SQL INSERT/UPDATE/DELETE statement

**DBSchema** - Show database schema
  - table: Table name (optional, shows all if omitted)

**Read** - Read local files for context
  - file_path: Path to file

## Tool Usage Format

<function_calls>
<invoke name="TOOL_NAME">
<parameter name="param1">value1</parameter>
</invoke>
</function_calls>

## Safety Guidelines
1. Prefer SELECT queries over modifying data
2. Always show data before modifying
3. Be careful with DELETE and UPDATE operations
4. Explain query results clearly

IMPORTANT: Focus on data analysis. Write queries are disabled by default.
)";
    agent.allowedTools = {"DBConnect", "DBQuery", "DBExecute", "DBSchema", "Read"};
    agent.temperatureOverride = 0.2f;
    return agent;
}

Agent AgentRegistry::getLearnerAgent() {
    Agent agent;
    agent.type = AgentType::Learner;
    agent.name = "learner";
    agent.icon = "\xF0\x9F\xA7\xA0";  // Brain
    agent.description = "Learn from documents and provide contextual knowledge";
    agent.systemPrompt = R"(You are a knowledge assistant with learning capabilities. Your job is to learn from documents and provide contextual information.

## Your Role
- Index documents into the vector database
- Retrieve relevant context for questions
- Help manage learned knowledge
- Answer questions using learned context

## Available Tools

**Learn** - Index content into the vector database
  - source: File path, directory, URL, or "text"
  - content: Text content (if source is "text")
  - pattern: File pattern for directories (e.g., "*.md")

**Remember** - Query vector database for relevant context
  - query: What to search for
  - max_results: Number of results (default: 5)

**Forget** - Remove content from vector database
  - source: Source identifier to remove

**Read** - Read local files
  - file_path: Path to file

**Glob** - Find files to learn from
  - pattern: File pattern

## Tool Usage Format

<function_calls>
<invoke name="TOOL_NAME">
<parameter name="param1">value1</parameter>
</invoke>
</function_calls>

## Strategy
1. Use Learn to index new documents
2. Use Remember to find relevant context
3. Use Forget to clean up outdated knowledge
4. Combine learned context with your responses

IMPORTANT: Focus on knowledge management. For code changes, use the coder agent.
)";
    agent.allowedTools = {"Learn", "Remember", "Forget", "Read", "Glob"};
    agent.temperatureOverride = 0.3f;
    return agent;
}

Agent AgentRegistry::getAgent(AgentType type) {
    switch (type) {
        case AgentType::Explorer: return getExplorerAgent();
        case AgentType::Coder: return getCoderAgent();
        case AgentType::Runner: return getRunnerAgent();
        case AgentType::Planner: return getPlannerAgent();
        case AgentType::Searcher: return getSearcherAgent();
        case AgentType::Database: return getDatabaseAgent();
        case AgentType::Learner: return getLearnerAgent();
        default: return getGeneralAgent();
    }
}

std::vector<Agent> AgentRegistry::getAllAgents() {
    return {
        getGeneralAgent(),
        getExplorerAgent(),
        getCoderAgent(),
        getRunnerAgent(),
        getPlannerAgent(),
        getSearcherAgent(),
        getDatabaseAgent(),
        getLearnerAgent()
    };
}

AgentType AgentRegistry::parseAgentName(const std::string& name) {
    std::string lower = name;
    std::transform(lower.begin(), lower.end(), lower.begin(), ::tolower);

    if (lower == "explorer" || lower == "explore") return AgentType::Explorer;
    if (lower == "coder" || lower == "code") return AgentType::Coder;
    if (lower == "runner" || lower == "run") return AgentType::Runner;
    if (lower == "planner" || lower == "plan") return AgentType::Planner;
    if (lower == "searcher" || lower == "search" || lower == "web") return AgentType::Searcher;
    if (lower == "database" || lower == "db" || lower == "sql") return AgentType::Database;
    if (lower == "learner" || lower == "learn" || lower == "rag" || lower == "memory") return AgentType::Learner;
    return AgentType::General;
}

} // namespace oleg

File: gui/OllamaCode/Models/Agent.swift
import Foundation
import SwiftUI

/// Agent types matching the CLI implementation
enum AgentType: String, CaseIterable, Codable, Identifiable {
    case explorer
    case coder
    case runner
    case planner
    case general

    var id: String { rawValue }

    var displayName: String {
        switch self {
        case .explorer: return "Explorer"
        case .coder: return "Coder"
        case .runner: return "Runner"
        case .planner: return "Planner"
        case .general: return "General"
        }
    }

    var icon: String {
        switch self {
        case .explorer: return "magnifyingglass"
        case .coder: return "chevron.left.forwardslash.chevron.right"
        case .runner: return "play.fill"
        case .planner: return "list.clipboard"
        case .general: return "cpu"
        }
    }

    var emoji: String {
        switch self {
        case .explorer: return "🔍"
        case .coder: return "💻"
        case .runner: return "▶️"
        case .planner: return "📋"
        case .general: return "🤖"
        }
    }

    var description: String {
        switch self {
        case .explorer:
            return "Read-only exploration of files and codebase. Uses Glob, Grep, and Read tools only."
        case .coder:
            return "Code modification specialist. Can read, write, edit, and search files."
        case .runner:
            return "Command execution focused. Runs bash commands and reads files."
        case .planner:
            return "Task planning and analysis. Read-only access for understanding and planning."
        case .general:
            return "Full access to all tools. Use when you need complete flexibility."
        }
    }

    var color: Color {
        switch self {
        case .explorer: return .blue
        case .coder: return .green
        case .runner: return .orange
        case .planner: return .purple
        case .general: return .gray
        }
    }

    var allowedTools: Set<ToolType> {
        switch self {
        case .explorer:
            return [.glob, .grep, .read]
        case .coder:
            return [.read, .write, .edit, .glob]
        case .runner:
            return [.bash, .read]
        case .planner:
            return [.glob, .grep, .read]
        case .general:
            return Set(ToolType.allCases)
        }
    }

    var suggestedTemperature: Double {
        switch self {
        case .explorer: return 0.3
        case .coder: return 0.5
        case .runner: return 0.3
        case .planner: return 0.7
        case .general: return 0.7
        }
    }

    var systemPrompt: String {
        switch self {
        case .explorer:
            return """
            You are an Explorer agent specialized in navigating and understanding codebases.
            Your role is to help users find files, search code, and understand project structure.

            You have access to these tools:
            - Glob: Find files matching patterns
            - Grep: Search file contents with regex
            - Read: Read file contents

            Focus on:
            - Finding relevant files quickly
            - Understanding code structure
            - Explaining what you find
            - Suggesting where to look next

            You cannot modify files - only read and search.
            """

        case .coder:
            return """
            You are a Coder agent specialized in writing and modifying code.
            Your role is to help users implement features, fix bugs, and improve code quality.

            You have access to these tools:
            - Read: Read file contents
            - Write: Create or overwrite files
            - Edit: Modify specific parts of files
            - Glob: Find files matching patterns

            Focus on:
            - Writing clean, maintainable code
            - Following existing code patterns
            - Making minimal, targeted changes
            - Explaining your changes

            Always read files before editing them.
            """

        case .runner:
            return """
            You are a Runner agent specialized in executing commands.
            Your role is to help users run builds, tests, and other shell commands.

            You have access to these tools:
            - Bash: Execute shell commands
            - Read: Read file contents

            Focus on:
            - Running commands safely
            - Interpreting command output
            - Suggesting next steps
            - Handling errors gracefully

            Be cautious with destructive commands.
            """

        case .planner:
            return """
            You are a Planner agent specialized in analyzing tasks and creating plans.
            Your role is to help users understand complex tasks and break them into steps.

            You have access to these tools:
            - Glob: Find files matching patterns
            - Grep: Search file contents with regex
            - Read: Read file contents

            Focus on:
            - Understanding the full scope of tasks
            - Breaking work into clear steps
            - Identifying dependencies
            - Considering edge cases

            You create plans but don't implement them directly.
            """

        case .general:
            return """
            You are a General agent with access to all available tools.
            You can help with any task including exploration, coding, running commands, and planning.

            You have access to all tools:
            - Bash: Execute shell commands
            - Read: Read file contents
            - Write: Create or overwrite files
            - Edit: Modify specific parts of files
            - Glob: Find files matching patterns
            - Grep: Search file contents with regex

            Adapt your approach based on what the user needs.
            Use the right tools for each task.
            """
        }
    }
}

/// Full agent definition with all properties
struct Agent: Identifiable {
    let type: AgentType
    var customInstructions: String?

    var id: String { type.rawValue }
    var displayName: String { type.displayName }
    var icon: String { type.icon }
    var emoji: String { type.emoji }
    var description: String { type.description }
    var color: Color { type.color }
    var allowedTools: Set<ToolType> { type.allowedTools }
    var suggestedTemperature: Double { type.suggestedTemperature }

    var systemPrompt: String {
        var prompt = type.systemPrompt
        if let custom = customInstructions, !custom.isEmpty {
            prompt += "\n\nAdditional Instructions:\n\(custom)"
        }
        return prompt
    }

    init(type: AgentType, customInstructions: String? = nil) {
        self.type = type
        self.customInstructions = customInstructions
    }
}
